File: KE_Python/Mandlebrot/mandlebrot.py
from math import floor

ASCII = " .:;+=xX$&"
ASCII_LEN_MINUS_ONE = len(ASCII) - 1

WIDTH = 100
HEIGHT = 70

RE_START = -2
RE_END = 1
IM_START = -1
IM_END = 1

MAX_ITERS = 100

if __name__ == "__main__":
    iters = []
    for y in range(HEIGHT):
        for x in range(WIDTH):
            real = (x / WIDTH) * (RE_END - RE_START) + RE_START
            imaginary = (y / HEIGHT) * (IM_END - IM_START) + IM_START

            base_complex_number = complex(real, imaginary)

            eval_cmplx_no = complex(0, 0)
            iter_count = 0
            while ((eval_cmplx_no.real * eval_cmplx_no.real) + (eval_cmplx_no.imag * eval_cmplx_no.imag)) < 4.0 and iter_count < MAX_ITERS:
                eval_cmplx_no = (eval_cmplx_no * eval_cmplx_no) + base_complex_number
                iter_count += 1

            iters.append(iter_count)

    max_from_list = max(iters)
    min_from_list = min(iters)
    diff = max_from_list - min_from_list

    print(f"Max is {max_from_list}, min is {min_from_list}, diff = {diff}.")

    s = ""
    row_pos = 0
    for iter_count in iters:
        if iter_count == MAX_ITERS:
            s += ASCII[ASCII_LEN_MINUS_ONE]
        else:
            mapped = iter_count - min_from_list
            pos = mapped / diff
            s += ASCII[floor(pos * ASCII_LEN_MINUS_ONE)]

        row_pos += 1
        if row_pos == WIDTH:
            row_pos = 0
            s += "\n"

    print(s)

File: KE_CPP/Mandlebrot/Mandlebrot.cpp
#include <cstdio>
#include <cstdlib>
#include <string>
#include <vector>


const char* ASCII = " .:;+=xX$&";
// int ASCII_LEN_MINUS_ONE = (sizeof(ASCII) / sizeof(*ASCII)) - 1; Gives pointer size rather than array size
const size_t ASCII_LEN_MINUS_ONE = 9;

const size_t WIDTH = 50;
const size_t HEIGHT = 35;

const float RE_START = -2.0;
const float RE_END = 1.0;
const float IM_START = -1.0;
const float IM_END = 1.0;

const uint8_t MAX_ITERS = 50;

int main()
{
    std::vector<uint8_t> iters;

    for (size_t y = 0; y < HEIGHT; y++) {
        for (size_t x = 0; x < WIDTH; x++) {
            float real = (((float)x / (float)WIDTH) * (RE_END - RE_START)) + RE_START;
            float imaginary = (((float)y / (float)HEIGHT) * (IM_END - IM_START)) + IM_START;

            float re = 0.0;
            float im = 0.0;
            uint8_t iter = 0;

            while (re * re + im * im < 4.0 && iter < MAX_ITERS) {
                re = re * re - im * im + real;
                im = re * im + re * im + imaginary;
                iter++;
            }

            iters.push_back(iter);
        }
    }


    uint8_t max = 0, min = MAX_ITERS;
    for (uint8_t iter : iters) {
        if (iter < MAX_ITERS && iter > max) {
            max = iter;
        }
        else if (iter < min) {
            min = iter;
        }
    }
    float diff = (float)(max - min);

    std::string s = "";
    size_t row_pos = 0;
    for (uint8_t iter : iters) {
        if (iter == MAX_ITERS) {
            s += ASCII[ASCII_LEN_MINUS_ONE];
        }
        else {
            uint8_t mapped = iter - min;
            float pos = (float)mapped / diff;
            s += ASCII[(size_t)(pos * (float)ASCII_LEN_MINUS_ONE)];
        }

        row_pos++;
        if (row_pos == WIDTH) {
            row_pos = 0;
            s += '\n';
        }
    }

    printf("%s", s.c_str());

}
